desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java

```diff
@@ -83,44 +85,113 @@ public void onAdded(Offer offer) {
+            offerItemsWithSameIdAndDifferentHash.forEach(oldOfferItem -> {
+                offerBookListItems.remove(oldOfferItem);
+                if (log.isDebugEnabled()) {
+                    log.debug("onAdded: Removed old offer {}\n"
+                                    + "\twith old payloadHash = {} from list.\n"
+                                    + "\tThis may make a subsequent onRemoved( {} ) call redundant.",
+                            offerId,
+                            oldOfferItem.getHashOfPayload() == null ? "null" : oldOfferItem.getHashOfPayload().getHex(),
+                            oldOfferItem.getOffer().getId());
+                }
+            });
+        }
+    }
+
+    public void removeOffer(Offer offer, P2PDataStorage.ByteArray hashOfPayload) {
         // Update state in case that that offer is used in the take offer screen, so it gets updated correctly
         offer.setState(Offer.State.REMOVED);
-
         offer.cancelAvailabilityRequest();
-        // We don't use the contains method as the equals method in Offer takes state and errorMessage into account.
-        Optional<OfferBookListItem> candidateToRemove = offerBookListItems.stream()
-                .filter(item -> item.getOffer().getId().equals(offer.getId()))
+
+        if (log.isDebugEnabled()) {
+            log.debug("onRemoved: id = {}\n"
+                            + "\twith payload-hash = {}",
+                    offer.getId(),
+                    hashOfPayload == null ? "null" : hashOfPayload.getHex());
+        }
+
+        // Find the removal candidate in the OfferBook list with matching offerId and payload-hash.
+        Optional<OfferBookListItem> candidateWithMatchingPayloadHash = offerBookListItems.stream()
+                .filter(item -> item.getOffer().getId().equals(offer.getId()) && (
+                        item.hashOfPayload == null
+                                || item.hashOfPayload.equals(hashOfPayload))
+                )
                 .findAny();
-        candidateToRemove.ifPresent(offerBookListItems::remove);
+
+        if (!candidateWithMatchingPayloadHash.isPresent()) {
+            if (log.isDebugEnabled()) {
+                log.debug("UI view list does not contain offer with id {} and payload-hash {}",
+                        offer.getId(),
+                        hashOfPayload == null ? "null" : hashOfPayload.getHex());
+            }
+            return;
+        }
+
+        OfferBookListItem candidate = candidateWithMatchingPayloadHash.get();
+
+        // Remove the candidate only if the candidate's offer payload hash matches the
+        // onRemoved hashOfPayload parameter.  We may receive add/remove messages out of
+        // order (from api's 'editoffer'), and use the offer payload hash to
+        // ensure we do not remove an edited offer immediately after it was added.
+        if ((candidate.getHashOfPayload() == null || candidate.getHashOfPayload().equals(hashOfPayload))) {
```

Comment:
How can there be candidates with null hash?

Reply:
There can be a null hash each time a user clicks on the view, and calling `OfferBook#fillOfferBookListItems()`.  That method loads the view using `offerBookService.getOffers()`, and the returned list elements are mapped to new `OfferBookListItem(offer)`s with a null hash code.

We discussed this on keybase, when you mentioned the idea of adding a `transient var hash` to OfferPayload.
At that time, I also mentioned that without the cached OfferPayload hash,  OfferBook line 208 might be 
        `.map(o -> new OfferBookListItem(o, get32ByteHashAsByteArray(o.getOfferPayload()))`

There is a comment about null hashes in `OfferBookListItem`:  _A null value indicates the item's payload hash has not been set by onAdded or onRemoved since the most recent OfferBook view refresh._

Reply:
Does this mean an item with null hash will be removed if only the id matches? Seems like a way to get around the hash checks.

Reply:
It is a way to get around the hash check if the list contains an item loaded by `OfferBook#fillOfferBookListItems()` (where items are created with null hashes).

If we do create a transient OfferPayload hash field, fillOfferBookListItems() can load the view with non-null hashes, and this null hash check can be removed.